Core & Internals: Change unnecessary RSE settings to RSE attributes #5464

joeldierkes opened the pull request:
The RSE database table is convoluted with unnecessary entries. Columns
like `continent` do not need to be defined here, they should rather be a
RSE attribute.

This PR does not change the exposed interface.

I added the properties I thought were important, maybe we should add more? (e.g. `ISP`?)
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

joeldierkes committed:
Core & Internals: Change unnecessary RSE settings to RSE attributes #5464

The RSE database table is convoluted with unnecessary entries. Columns
like `continent` do not need to be defined here, they should rather be a
RSE attribute.

joeldierkes commented:
Since this is my first PR where I migrate data in the database, please review extra careful! :sweat_smile:

mlassnig commented on lib/rucio/core/rse.py:
```diff
@@ -161,6 +158,9 @@ def bulk_load(rse_datas: "Sequence[RseData]", load_name=False, load_columns=Fals
+
+
 @transactional_session
 def add_rse(rse, vo='def', deterministic=True, volatile=False, city=None, region_code=None, country_name=None, continent=None, time_zone=None,
```
I think we can actually update the signature as well. It needs to be fixed here as well though: https://github.com/rucio/probes/blob/master/atlas/check_sync_rses_with_cric

joeldierkes replied:
This will involve a bigger change (Rest/Client Api with deprecation, other places where `add_rse(` is called), I'd like to do that in a follow-up PR once the data is migrated.

mlassnig commented on lib/rucio/core/rse.py:
```diff
@@ -188,9 +188,8 @@ def add_rse(rse, vo='def', deterministic=True, volatile=False, city=None, region
-                         region_code=region_code, country_name=country_name,
-                         continent=continent, time_zone=time_zone, staging_area=staging_area, ISP=ISP, availability=availability,
+    new_rse = models.RSE(rse=rse, vo=vo, deterministic=deterministic, volatile=volatile,
+                         staging_area=staging_area, ISP=ISP, availability=availability,
```
don't think ISP is in any way useful

mlassnig commented on lib/rucio/core/rse.py:
```diff
@@ -1400,7 +1423,7 @@ def update_rse(rse_id: str, parameters: 'Dict[str, Any]', session=None):
 
     :param rse_id: the id of the new rse.
-    :param  parameters: A dictionnary with property (name, read, write, delete as keys).
+    :param parameters: A dictionnary with property (name, read, write, delete as keys).
```
dictionnary -> dictionary

mlassnig commented on lib/rucio/db/sqla/migrate_repo/versions/2190e703eb6e_move_rse_settings_to_rse_attributes.py:
```diff
@@ -0,0 +1,113 @@
+    '''
+
+    if context.get_context().dialect.name in ['oracle', 'mysql', 'postgresql']:
+        conn = get_bind()
```
probably needs a transaction

joeldierkes replied:
Same as https://github.com/rucio/rucio/pull/5665#discussion_r907260030

mlassnig commented on lib/rucio/db/sqla/models.py:
```diff
@@ -752,11 +752,6 @@ class RSE(BASE, SoftModelBase):
-    country_name = Column(String(255))
-    continent = Column(String(2))
-    time_zone = Column(String(255))
     ISP = Column(String(255))
```
i think these two (ISP and ASN) can go as well

joeldierkes replied:
Erased from existence

joeldierkes commented:
Version two of this PR contains some changes:

Same as #5664, the columns do not get removed from the database for backwards compatibility. The `add_rse`, `update_rse` and `{get,list}_rses` use the new behavior.

This PR therefore needs a follow-up: remove the database columns and backwards compatibility. I'd include the signature change there, since this involves more changes. In this PR I want to focus on the data migration/new behavior.

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/2190e703eb6e_move_rse_settings_to_rse_attributes.py:
```diff
@@ -0,0 +1,111 @@
+                schema=SCHEMA
+            )
+
+            rses = conn.execute(rse_table.select().where(sa.sql.column(setting) != None)).fetchall()  # noqa: E711
```
As discussed on Slack already, I am not sure this works. 
One key principle is that you can also do a `alembic upgrade head --sql` which generates you all the SQL statements (to verify and execute by hand). But if I do this with this revision, it fails.

joeldierkes replied:
Yes, I changed it as discussed.

The sql query used to downgrade the db is using a subquery since oracle does not support the update query with multiple tables.

joeldierkes committed:
Core & Internals: Allow updating the `continent` attribute of a RSE #5464

The `continent` associated attribute is not updatable. Since all other
related attributes (e.g. `city`, `country_name`) can be updated, we
should add this one to make it consistent.
